Openmv and MicroPython tutorial (#94)
arduino/arduino-pro-content: content/tutorials/portenta-h7/por-openmv-gs/content.md

# Getting Started with OpenMV and MicroPython

## Overview

> In this tutorial, you will learn about some of the basic features of the OpenMV editor and run Micropython script that will control the PWM on the I/O pins of the Portenta. 

Reviewer: ```suggestion
In this tutorial, you will learn about some of the basic features of the OpenMV IDE and run a MicroPython script that controls the PWM on the I/O pins of the Arduino Portenta H7. 
```

### You Will Learn

> - How to use the OpenMV IDE

Reviewer: ```suggestion
- The basic features of the OpenMV IDE
```

- How to use the OpenMV IDE to run MicroPython on Portenta
- How to use PWM and I/O pins on the Portenta

### Required Hardware and Software
- Portenta H7 board (<https://store.arduino.cc/portenta-h7>)
- Arduino Portenta Vision Shield (https://store.arduino.cc/portenta-vision-shield)
- USB C cable (either USB A to USB C or USB C to USB C)
- Arduino IDE 1.8.10+  or Arduino Pro IDE 0.0.4+ 
- Portenta Bootloader Version 20+
- OpenMV IDE 2.6.4+

> ## Portenta and the OpenMV IDE

Reviewer 2: I would remove the part about what the founders have to say, pulse width modulation and the I/O pins. I would instead summarise those to give the users an idea/ overview of what they can do. Focusing more on how to control the PWM part

The OpenMV IDE was built for Machine Vision applications. It is meant to provide an Arduino like experience for simple computer vision tasks using a camera sensor. OpenMV comes with its own firmware that is built on MicroPython. Among other hardware it supports the Portenta board.

This is where OpenMV comes in. [Here](https://openmv.io/) you can read more about the OpenMV IDE.

## MicroPython and PWM

> With OpenMV and the Portenta board, it is possible to use MicroPython in scripts. MicroPython includes a lot of classes and libraries that makes it easier for us to use the Portenta to its full potential.

Reviewer: ```suggestion
With OpenMV and the Portenta board, it is possible to run MicroPython scripts. MicroPython includes a lot of classes and libraries that makes it easier for us to use the Portenta to its full potential.
```

***IMPORTANT: Before you connect the Portenta to the OpenMV IDE make sure you update the bootloader as explained in the "Flashing the OpenMV Firmware" section!***

### 1. Downloading the OpenMV IDE
Open the [OpenMV download](https://openmv.io/pages/download) page in your browser and download the version that you need for your operating system. Alternatively, you may use the following direct download links of the OpenMV IDE 2.6.5:

- [For Windows Xp, Vista, 7, 8, 10 or Later](https://github.com/openmv/openmv-ide/releases/download/v2.6.5/openmv-ide-windows-2.6.5.exe)
- [For OsX Snow Leopard 10.6 or Later](https://github.com/openmv/openmv-ide/releases/download/v2.6.5/openmv-ide-mac-2.6.5.dmg)
- [For Ubuntu 12.04 Lts 32-Bit or Later](https://github.com/openmv/openmv-ide/releases/download/v2.6.5/openmv-ide-linux-x86-2.6.5.run)
- [For Ubuntu 12.04 Lts 64-Bit or Later](https://github.com/openmv/openmv-ide/releases/download/v2.6.5/openmv-ide-linux-x86_64-2.6.5.run)
- [For Raspberry Pi 0, 1, 2, 3, and 4 or Later](https://github.com/openmv/openmv-ide/releases/download/v2.6.5/openmv-ide-linux-arm-2.6.5.tar.gz)

> Follow the instructions of the installer.

Reviewer 2: Move this line to the first sentence

### 2. Flashing the OpenMV Firmware

Connect the Portenta to your computer via the USB-C cable if you haven't done so yet. Make sure you first update the bootloader to the latest version using the *PortentaH7_updateBootloader* sketch in the examples menu in the Arduino IDE. 

> Instructions on how to update the bootloader can be found in the ["Updating the Portenta Bootloader" tutorial](https://www.arduino.cc/pro/tutorials/portenta-h7/por-ard-bl).

Reviewer 2: This can be made as a note

> ***In bootloader versions 17 and older there was a bug that could put the Portenta in a boot loop when the transmission aborted while flashing a large firmware file. This was fixed in the bootloader version 18. We strongly advise to update the bootloader before you proceed with the next step of this tutorial!***

Reviewer 2: I would add this as a troubleshooting or move it to he bootloader tutorial itself

![Installing firmware on portenta board in OpenMV](assets/por_openmv_firmware_update.png)

>Installing the OpenMV firmware will overwrite any existing sketches in the internal flash of Portenta. As a result the M7 port won't be exposed in the Arduino IDE anymore. To re-flash the M7 with an Arduino firmware you need to put the board into bootloader mode. To do so double press the reset button on the Portenta H7 board. The built-in green LED will start fading in and out. In bootloader mode you will see the Portenta M7 port again in the Arduino IDE.

The Portenta will start flashing its blue LED when it's ready to be connected. After confirming the completion dialog the Portenta should already be connected to the OpenMV IDE, otherwise click the "connect" button once again.

![When the Portenta is successfully connected to the OpenMV IDE a green play button appears in the lower left](assets/por_openmv_board_connected.png)

> ### 3. Wiring the LED for the Scrip

Reviewer: ```suggestion
### 3. Wiring the LED for the Script
```

A pop-up will ask you how you would like to proceed "DFU bootloader(s) found. What would you like to do?". Select "Reset Firmware to Release Version". This will install the latest OpenMV firmware on the Portenta. If it asks you whether it should erase the internal file system you can click "No".

![Install the latest version of the OpenMV firmware](assets/por_openmv_reset_firmware.png)

Portenta's green LED will start flashing while the OpenMV firmware is being uploaded to the board. A terminal window will open which shows you the upload progress. Wait until the green LED stops flashing and fading. You will see a message saying "DFU firmware update complete!" when the process is done.

> 

Reviewer 2: Would be nice to have an image here to support the explanation

> ***Installing the OpenMV firmware will overwrite any existing sketches in the internal flash of Portenta. As a result the M7 port won't be exposed in the Arduino IDE anymore. To re-flash the M7 with an Arduino firmware you need to put the board into bootloader mode. To do so double press the reset button on the Portenta H7 board. The built-in green LED will start fading in and out. In bootloader mode you will see the Portenta M7 port again in the Arduino IDE.***

Reviewer 2: This can be part of the main step. need not be a note

The Portenta will start flashing its blue LED when it's ready to be connected. After confirming the completion dialog the Portenta should already be connected to the OpenMV IDE, otherwise click the "connect" button once again.

![When the Portenta is successfully connected to the OpenMV IDE a green play button appears in the lower left](assets/por_openmv_board_connected.png)

> ### 3. Wiring the LED to the Portenta

Reviewer 2: We arent wiring anything to the board per say. 
This can be instead "Controlling the LED's Intensity"

```py
import pyb # Import module for board related functions
import time # Import module for board related functions
from pyb import Pin, Timer # Import module for board related functions
```

A module in Python is a confined bundle of functionality. By importing it into the script it gets made available.

In order to control a LED connected to one of our pins, we first need to configure the pins behavior. After, we can set up the PWM.

> In the `pin1` variable we define what pin on the board we intend to use. We also define if it should be an input or output pin with `Pin.OUT_PP`. Lastly, we can set the behavior of the pull up resistor on the pin, for now we set it as `Pin.PULL_NONE`.

Reviewer: ```suggestion
In the `pin1` variable we define what pin on the board we intend to use. We also define if it should be an input or output pin with `Pin.OUT_PP`. Lastly, we can set the behaviour of the pull-up resistor on the pin, for now we set it as `Pin.PULL_NONE`.
```

> >For more info about the options available, please see [Here](https://docs.micropython.org/en/latest/library/pyb.Pin.html?)

Reviewer: ```suggestion
For more info about the options available, please see [here](https://docs.micropython.org/en/latest/library/pyb.Pin.html)
```

> [Illustration of a LEDs positive leg on a breadboard wired to the PC6 pin on the Portenta, and the negative end going to the GND pin on the Portenta]()

Reviewer 2: Is this an image or a link ?

Author: The illustration we talked about should be here.

```py
import pyb # Import module for board related functions
import time # Import module for board related functions
from pyb import Pin, Timer # Import module for board related functions
```

A module in Python is a confined bundle of functionality. By importing it into the script it gets made available.

> ### 5. Preparing the Pin and PWM

Reviewer 2: Step 4 is too short. Try to combine it with Step 5

channel1 = timer1.channel(1, Timer.PWM, pin=pin1, pulse_width=0)
```

We also enter parameters for how much the LED will change each time the PWM pulses with the `step` variable, what value it should start on with `minWidth` and 
when it will reset with the `maxWidth` variable.

```py
# maximum and minimum pulse-width, corresponds to maximum and minimum brightness
maxWidth = 5000
minWidth = 0

# How much to change the pulse-width by each step
step = 500
# Set starting value

> curWidth = minWidth

Reviewer: ```suggestion
currentWidth = minWidth
```

curWidth = minWidth
```

> In the `pin1` variable we define what pin on the board we intend to use. We also define if it should be an input or output pin with `Pin.OUT_PP`. Lastly, we can set the behavior of the pull up resistor on the pin, for now we set it as `Pin.PULL_NONE`.

Reviewer 2: Its a good practice to give the explanation first and then the code. here you are writing the code first followed by the explanation. As a user I wouldnt understand why am i including certain variables

>For more info about the options available, please see [Here](https://docs.micropython.org/en/latest/library/pyb.Pin.html?)

> With the `timer1` variable we will determine what id the timer will have and what frequency it will use. With a timer frequency of 1000 Hz, each cycle takes 1 millisecond. When setting up the PWM channel we can then use these variables. With the `timer1.channel()` function we can customize the PWM channels behavior. `pulse_width` sets the initial pulse width for the PWM.

Reviewer 2: You could also split the code while explaining. Something more like 
Explanation > Code > Explanation > Code ....

We also enter parameters for how much the LED will change each time the PWM pulses with the `step` variable, what value it should start on with `minWidth` and when it will reset with the `maxWidth` variable.

> ### 6. Using MicroPython

Reviewer 2: This title doesnt make sense

Putting our code inside a `while True:` function will make the code run in a loop.

```py
while True:

  channel1.pulse_width(curWidth)

  # this determines how often we change the pulse-width.
  pyb.delay(500)

  curWidth = curWidth + step

  if curWidth > maxWidth:
    print("Max width reached!")
    curWidth = minWidth
```

We use the `channel1.pulse_width()` function to set the pulse width of the PWM channel, changing the intensity of the LED. Using `pyb.delay()` we can make the sketch stop running for the determined amount of time, which is 500 milliseconds in this example. At the end, we check if the value has reached our defined end, we print a message to the serial terminal and reset the `curWidth` variable.

### 7. Uploading the Script

If the wiring is correct the LED light should start incrementally getting brighter before resetting and starting over. Printing a message when the limit has been reached. 

> 

Reviewer 2: Would be nice to move the complete sketch at the end of the previous step

maxWidth = 5000
minWidth = 0

# How much to change the pulse-width by each step
step = 500
# Set starting value
curWidth = minWidth

while True:
  channel1.pulse_width(curWidth)

  # How often the pulse-width is changed
  pyb.delay(500)
  curWidth = curWidth + step

  if curWidth > maxWidth:
    print("Max width reached!")
    curWidth = minWidth
```

To upload the code to your board in OpenMV, press the green play button in the lower left corner.

![Press the green play button to upload the script](assets/por_openmv_board_connected.png)

If the wiring is correct the LED light should start incrementally getting brighter before resetting and starting over. Printing a message when the limit has been reached. If everything went correctly you should get the same output in the serial terminal as seen below.

![Screenshot of openmv showing what will print in the serial terminal](assets/por_openmv_ide_result.png)

## Conclusion

In this tutorial, you learned how to use the OpenMV IDE to develop MicroPython scripts that run on the Portenta board. You learned how to use and configure I/O pins and PWM on the Portenta. Lastly, you learned how to use these to control a LED light.

> ### Next Steps

Reviewer: @BenjaminDannegard This section is rather meant to describe next steps based on what they learned here. What else could they explore that is related to this? I'd avoid referring to the other tutorials unless they are an actual sequel to this one.

Author: Alright I will change that section

> 

Reviewer 2: You can also add a screenshot of the output here
